trajopt_ifopt/include/trajopt_ifopt/constraints/cartesian_line_constraint.h

```diff
@@ -0,0 +1,193 @@
+ * @par
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef TRAJOPT_IFOPT_CARTESIAN_LINE_CONSTRAINT_H
+#define TRAJOPT_IFOPT_CARTESIAN_LINE_CONSTRAINT_H
+
+#include <trajopt_utils/macros.h>
+TRAJOPT_IGNORE_WARNINGS_PUSH
+#include <Eigen/Eigen>
+#include <ifopt/constraint_set.h>
+
+#include <tesseract_kinematics/core/forward_kinematics.h>
+#include <tesseract_environment/core/environment.h>
+#include <tesseract_environment/core/utils.h>
+TRAJOPT_IGNORE_WARNINGS_POP
+
+#include <trajopt/utils.hpp>
+#include <trajopt_ifopt/variable_sets/joint_position_variable.h>
+
+namespace trajopt
+{
+/**
+ * @brief Contains kinematic information for the cartesian position cost; inlude cart point .h & remove?
+ */
+struct CartPosKinematicInfo
+{
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  using Ptr = std::shared_ptr<CartPosKinematicInfo>;
+  using ConstPtr = std::shared_ptr<const CartPosKinematicInfo>;
+
+  CartPosKinematicInfo() = default;
+  CartPosKinematicInfo(tesseract_kinematics::ForwardKinematics::ConstPtr manip,
+                       tesseract_environment::AdjacencyMap::ConstPtr adjacency_map,
+                       const Eigen::Isometry3d& world_to_base,
+                       std::string link,
+                       const Eigen::Isometry3d& tcp = Eigen::Isometry3d::Identity())
+    : manip(std::move(manip))
+    , adjacency_map(std::move(adjacency_map))
+    , world_to_base(world_to_base)
+    , link(std::move(link))
+    , tcp(tcp)
+  {
+    this->kin_link = this->adjacency_map->getLinkMapping(this->link);
+    if (this->kin_link == nullptr)
+    {
+      CONSOLE_BRIDGE_logError("Link name '%s' provided does not exist.", this->link.c_str());
+      assert(false);
+    }
+  }
+
+  tesseract_kinematics::ForwardKinematics::ConstPtr manip;
+  tesseract_environment::AdjacencyMap::ConstPtr adjacency_map;
+  Eigen::Isometry3d world_to_base;
+  std::string link;
+  tesseract_environment::AdjacencyMapPair::ConstPtr kin_link;
+  Eigen::Isometry3d tcp;
+};
+
+class CartLineConstraint : public ifopt::ConstraintSet
+{
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  using Ptr = std::shared_ptr<CartLineConstraint>;
+  using ConstPtr = std::shared_ptr<const CartLineConstraint>;
+
+  CartLineConstraint(const Eigen::Isometry3d& origin_pose,
+                     const Eigen::Isometry3d& target_pose,
+                    CartPosKinematicInfo::ConstPtr kinematic_info,
+                    JointPosition::ConstPtr position_var,
```

Comment:
Not asking you to do this, but we might want to think about if this error can benefit from vectorization in any way. For example how joint velocity takes a vector of vars and uses Eigen for the math. This constraint will most likely be used on a vector of variables since it is a path constraint. In the future we may need to rethink our variable structure since I'm not sure the overhead involved with accessing these one at a time.